x-pack/plugins/licensing/server/license_fetcher.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type * as estypes from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import { createHash } from 'crypto';
+import stringify from 'json-stable-stringify';
+import type { MaybePromise } from '@kbn/utility-types';
+import { isPromise } from '@kbn/std';
+import type { IClusterClient, Logger } from '@kbn/core/server';
+import type {
+  ILicense,
+  PublicLicense,
+  PublicFeatures,
+  LicenseType,
+  LicenseStatus,
+} from '../common/types';
+import { License } from '../common/license';
+import type { ElasticsearchError, LicenseFetcher } from './types';
+
+export const getLicenseFetcher = ({
+  clusterClient,
+  logger,
+  cacheDurationMs,
+}: {
+  clusterClient: MaybePromise<IClusterClient>;
+  logger: Logger;
+  cacheDurationMs: number;
+}): LicenseFetcher => {
+  let currentLicense: ILicense | undefined;
+  let lastSuccessfulFetchTime: number | undefined;
+
+  return async () => {
+    const client = isPromise(clusterClient) ? await clusterClient : clusterClient;
+    try {
+      const response = await client.asInternalUser.xpack.info(undefined, { maxRetries: 3 });
```

Comment:
Just noticed `maxRetries: 3` wasn't part of the old code, so IIUC we're now asking `elasticsearch-js` to retry 3 times in a row before actually reporting error 👍🏼 

That's already an improvement, but I wonder if we could have a more resilient strategy using _exponential backoff_ and perhaps a few more retries (5?). That could probably cover https://github.com/elastic/sdh-kibana/issues/4194 already.

I also agree that ES probably shouldn't report itself as available if the `GET /_xpack` API is not. Will try to investigate/confirm if that's the case.

Reply:
> Just noticed maxRetries: 3 wasn't part of the old code, so IIUC we're now asking elasticsearch-js to retry 3 times in a row before actually reporting error 👍🏼

It's actually the default value of the option. I'm not sure why I added it tbh, I think I wanted it to be more explicit. I should just remove it for now.